// vscode/index.js
const path = require("node:path");

const controllers = [];

class Range {
  constructor(startLine, startCharacter, endLine, endCharacter) {
    this.start = { line: startLine, character: startCharacter };
    this.end = { line: endLine, character: endCharacter };
  }
}

class Uri {
  constructor(fsPath) {
    this.fsPath = path.normalize(fsPath);
  }

  static file(fsPath) {
    return new Uri(fsPath);
  }
}

class FakeCommentThread {
  constructor(uri, range, comments) {
    this.uri = uri;
    this.range = range;
    this.comments = comments;
    this.canReply = true;
    this.contextValue = undefined;
    this.label = undefined;
    this.state = undefined;
    this.disposed = false;
  }

  dispose() {
    this.disposed = true;
  }
}

class FakeCommentController {
  constructor(id, label) {
    this.id = id;
    this.label = label;
    this.createdThreads = [];
    this.disposed = false;
  }

  createCommentThread(uri, range, comments) {
    const thread = new FakeCommentThread(uri, range, comments);
    this.createdThreads.push(thread);
    return thread;
  }

  dispose() {
    this.disposed = true;
  }
}

const comments = {
  createCommentController(id, label) {
    const controller = new FakeCommentController(id, label);
    controllers.push(controller);
    return controller;
  },
};

const CommentThreadState = {
  Resolved: 0,
  Unresolved: 1,
};

const CommentMode = {
  Preview: 0,
};

function __getControllers() {
  return [...controllers];
}

function __reset() {
  controllers.length = 0;
}

module.exports = {
  __getControllers,
  __reset,
  comments,
  CommentMode,
  CommentThreadState,
  Range,
  Uri,
};
